x/ibc/23-commitment/types/merkle.go

```diff
@@ -155,32 +182,139 @@ func (MerkleProof) GetCommitmentType() exported.Type {
 
 // VerifyMembership verifies the membership pf a merkle proof against the given root, path, and value.
 func (proof MerkleProof) VerifyMembership(root exported.Root, path exported.Path, value []byte) error {
-	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() || len(value) == 0 {
-		return errors.New("empty params or proof")
+	if proof.ValidateBasic() != nil || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() || len(value) == 0 {
+		return sdkerrors.Wrap(ErrInvalidProof, "empty params or proof")
+	}
+
+	mpath, ok := path.(MerklePath)
+	if !ok {
+		return sdkerrors.Wrap(ErrInvalidProof, "path is not a merkle path for a merkle proof")
+	}
+
+	if len(proof.Proofs) == len(mpath.KeyPaths) {
+		return sdkerrors.Wrapf(ErrInvalidProof, "invalid chained proof. proof chain length %d not the same as path length %d", len(proof.Proofs), len(mpath.KeyPaths))
 	}
 
-	runtime := rootmulti.DefaultProofRuntime()
-	return runtime.VerifyValue(proof.Proof, root.GetHash(), path.String(), value)
+	for i, p := range proof.Proofs {
+		// While the proofs go from lowest subtree to the final tree, the path is defined from the root
+		// down to the leaf. Thus, we must pass in subpaths in reverse order during chained proof verification
+		subpath := []byte(mpath.KeyPaths[len(mpath.KeyPaths)-i-1].String())
+		existProof, ok := p.Proof.(*ics23.CommitmentProof_Exist)
+		if !ok {
+			return sdkerrors.Wrap(ErrInvalidProof, "proof is not an existence proof")
+		}
+		// For subtree verification, we simply calculate the root from the proof and use it to prove
+		// against the value
+		subroot, err := existProof.Exist.Calculate()
+		if err != nil {
+			return sdkerrors.Wrap(ErrInvalidProof, err.Error())
+		}
+		if i != len(proof.Proofs)-1 {
+			if !ics23.VerifyMembership(proof.Specs[i], subroot, p, subpath, value) {
+				return sdkerrors.Wrapf(ErrInvalidProof, "invalid proof for path: %s", path.String())
+			}
+		} else {
+			// For the final verification, we prove inclusion against the root that was passed into function
+			// rather than calculating subroot in order to verify that the given value was committed to by
+			// the given root
+			if !ics23.VerifyMembership(proof.Specs[i], root.GetHash(), p, subpath, value) {
+				return sdkerrors.Wrapf(ErrInvalidProof, "invalid proof for path: %s", path.String())
+			}
+		}
+		// Each subtree root becomes the proving value for the next proof in the chaining process
+		value = subroot
+	}
+	return nil
 }
 
 // VerifyNonMembership verifies the absence of a merkle proof against the given root and path.
+// VerifyNonMembership verifies a chained proof where the absence of a given path is proven
+// at the lowest subtree and then each subtree's inclusion is proved up to the final root.
 func (proof MerkleProof) VerifyNonMembership(root exported.Root, path exported.Path) error {
```

Comment:
Currently VerifyNonMembership will verify absence at the smallest subtree, and then prove inclusion of all subtree upto the root.

Thus in the multistore example, we can prove absence of a value in the ibc store, but we can't use this to prove that `storeABC` does not exist in multistore

That seems sufficient for all ibc usecases.

Reply:
That's fine, let's just make sure we panic if we ever try to check a proof of absence of `storeABC`